README
 -----------
| Changelog |
 -----------

May 22, 2019 - Max

 o Making Galax compile with Ocaml 4.07 and most recent library dependencies. Installation prerequisites: make, libpcre, opam, opam packages: {pxp, camomile, pcre, camlidl, num}.

February 1, 2012 - Jerome

 o Constructors:
    - Fixed bug with relative base URI in element constructors.

January 25, 2012 - Jerome

 o Parsing:
    - Fixed lexing issue with tokens after '/' that can only be in the
      operator state.

 o Constructors:
    - Fixed some error cases when xmlns/xml names are used in
      attribute or element constructors.

January 24, 2012 - Jerome

 o Parsing:
    - Added collation support in order-by clause.

January 23, 2012 - Jerome

 o Rewritings:
    - Fixed somewhat embarrassing bug on rewriting of overloaded
      functions, not properly checking for all cases in union types.
    - Fixed rewriting bug for overloaded function applied to anyURI,
      revealed by the previous fix.

January 20, 2012 - Jerome

 o Parsing:
    - Fixed a few regressions introduced by the latest change to the parser.

 o Streaming:
    - Fixed an issue with code-selection for global variables,
      streaming too aggressively.

 o Type-checking:
    - Fixed a bug in type intersection involving node().

January 19, 2012 - Jerome

 o Parsing:
    - Removing obolete rules leading to (harmless) shift/reduce.
    - Cleaned up some unused tokens and lexing rules.

January 18, 2012 - Jerome

 o F&O:
    - Final fix to fn:subsequence, issue with NaN or INF as parameter.
    - Fix to negative marker on fn:round resulting in -0 rather than 0.

January 17, 2012 - Jerome

 o F&O:
    - Fixed rounding issue with fn:subsequence.

January 10, 2012 - Jerome

 o Configuration:
    - Fixed bug in configuration generation introduced in previous patch.

 o F&O:
    - fn:doc-available returns false for the empty sequence, not empty.

January 10, 2012 - Jerome

 o Lexing:
    - Fixed lexical state issue arising from function types.
    - Added -print-states [on/off] option to the top-level tokenizer.

January 7, 2012 - Jerome

 o Configuration:
    - Fixed lexing tool script for MacOSX.

January 6, 2012 - Jerome

 o Lexing:
    - Added a hook to the tokenizer from the command line, returning
      the set of lexing tokens, along with the lexical state's stack.

      Usage:
        glx tokenize [files]

January 5, 2012 - Jerome

 o Lexing:
    - Removed Xmldecl_lexer, unused.

 o Parsing:
    - Computed processing instruction accepts no expression content.
    - Direct processing instruction constructor checks for space after
      NCName.
    - Computed processing instruction names, should make sure to cast
      to xs:NCName. (Note: cast to xs:NCName does not properly check
      for facet constraints at this point).

 o Overloaded functions:
    - Fixed overloading of unary arithmetic operators over the empty
      sequence.

 o Date and time:
    - Fixed rounding error when multiplying over a yearMonthDuration.

 o Functions and operators:
    - Added missing (new?) fn:doc-available, simply reusing fn:doc.

January 4, 2012 - Jerome

 o Documentation:
    - Some cleanup and basic update of the documentation.

 o Configuration:
    - Made the ./configure script more robust to variations in ocaml's
      version number.

 o Installation:
     - Fixed installed makefile for update usecases.

January 3, 2012 - Jerome

 o Parsing:
    - Fixed minor issue with new-line normalization.
    - Added strict conformance flag. Sample rule is no multiple
      expressions in the main module.
    - Fixed bug in parsing or processing instruction kind tests
      provided as a string, succeedings even when not a validNCName.
    - Fixed bug in parsing for attribute and schema-attribute node
      kinds.

December 21, 2011 - Jerome

  o Rewriting:
     - Fixed slightly too aggressive rewriting rules for
       fs:untyped-to-double, leading to errors in some cases.

December 20, 2011 - Jerome

  o Normalization:
     - Fixed odd bug on leading '/' and leading '//' not properly
       checking for document node.

  o Testing:
     - Simple switch of XQTS to v103
     - Added some missing test groups (some new namespace, errors, XML
       1.1 tests)

  o Parsing:
     - Fixed bug in parsing of default typeswitch clause with a variable.

  o XPath:
     - Fixed a couple of issues with reverse axis and
       following/preceding sibling (empty for attributes nodes).

  o Configuration:
     - Cleaned up a few wrong version references and links to Web site
       in the code.

December 16, 2011 - Jerome

  o Rewriting:
     - Reverted slightly too aggressive rewriting rule for casting,
       introducing a regression.
     - Consolidated E1[E2] rewrite, making sure they are correct and
       do not introduce regressions.

December 15, 2011 - Jerome

  o Rewriting:
     - Fixed some of the basic rewrites for E1[E2] when E2 is a
       numeric, leads to better access times for those cases.

December 8, 2011 - Jerome

  o Testing:
     - Fixed serialization of XML optimized plans during testing
     - Some fixes to the GLX tests catalog

December 8, 2011 - Jerome

  o Streaming:
     - Further cleanup of namespace resolution.

December 6, 2011 - Jerome

  o GIT:
     - Removed some files that shouldn't be tracked.
  o Streaming:
     - Slightly neater attribute resolution code.

December 6, 2011 - Jerome

  o Streaming:
     - Cleaned up some of the handling of namespace attributes.

December 6, 2011 - Jerome

  o Streaming:
     - Reduces namespace environment manipulation during stream
       resolution, resulting in much faster processing in some cases.

December 5, 2011 - Jerome

  o Build:
      - Now builds on ocaml 3.12.1+rc1
      - Upgraded all the major libraries:
      	   - findlib 1.2.7
           - pcre-ocaml 6.2.3
           - pxp 1.2.2
           - camomile 0.8.3

  o Data model:
      - Major revision of the streams support, now relying on a single
        stream type, in an attempt to reduce the number of events
        being copied.

March 30, 2011 - Jerome

  o String matching:
     - Fixed problem with tokenization of empty sequence, introduced
       by latest fix.

March 9, 2011 - Jerome

  o Build:
      - Now builds on ocaml 3.12, with camomile 0.8

  o String matching:
     - Fixed serious bug in fn:tokenize (failing for consecutive
        delimiters without text in between).

June 7, 2010 - Jerome

  o Errors:
      - Applied patch fixing typos in error messages. [Contributed by
        Stephane Glondu]

May 11, 2010 - Jerome

  o Lexing:
      - Improved efficiency of keywords lexing.

  o Usecases:
      - Updates usecases now points to the right local executable.

April 19, 2010 - Jerome

  o Modules:
      - Improved propagation of errors during module import.

  o Usecases:
      - Cleaned up use case execution and installation.

February 9, 2010 - Jerome

  o Toplevel:
      - Made sure command-line help prints out details for their
        respective sub-commands. 'glx help <subcommand>' and 'glx-map
        help <subcommand>' now dispatch properly to 'glx <subcommand>
        -help' and 'glx-map <subcommand> -help' respectively.
      - Fixed 'glx -version' and 'glx-map -version'.

January 25, 2010 - Jerome

  o Toplevel:
      - Migrated xquery2soap to the glx-map command line tool.
      - Reorganized the remaining command-line executable production
        in the main Makefile for clarity.

January 22, 2010 - Jerome

  o Toplevel:
      - Migrated Xquery2xquery, Wsdl2xquery, Xquery2plan, xmlplan2plan
        command lines tools to the glx-map command line tool.
      - Migrated options for WSDL mapping to the generic option
        support.

January 13, 2010 - Jerome

  o Regression testing:
      - Fixed the makefiles for regression testing, not using the new
        command-line options.

  o Toplevel:
      - Removed obsolete galax-mapschema.ml file.

January 13, 2010 - Jerome

  o Examples: More clean up of the Caml API example.
      - Fixed the Caml API tests to use the proper result comparison
        target.
      - Made sure the install / compile / run / clean works properly.

January 12, 2010 - Jerome

  o Examples:
      - Fixed installed Makefile.galax to make sure the caml_api
        examples compile properly. (Closes #2898247)
  o Documentation:
      - Started updating the documentation to reflect the new command
        line tools. (Related #2928716)
      - Removed unrelated JoinPlan.txt file in the documentation.

January 5, 2010 - Jerome

  o Toplevel: More clean-ups of the command-line tools.
      - galax --> glx
      - New tool focused on mappings: glx-map
      - galaxd --> glxd
      - galax-test --> glx-test
      - Fixed some module names with '-'.
      - Added help message in case no specific directive is passed to
        the command line-tool. e.g., glx help ; glx-map help
      - Renumbered some of the TARGETS in the main Makefile.

December 15, 2009 - Jerome

  o Toplevel:
      - Added projection (untested and likely not working) to main
          command-line tool: galax-project [options] --> galax project [options]

December 8, 2009 - Jerome

  o Cleanup: Reviewed unused/uncompiled modules.
      - Removed the unused Xquery_algebra_ast_annotation AST.
      - Removed Glx_jabber from the target list (unmaintained extension).
      - Made sure Factorize_global is compiling, and linked-in properly.
      - Removed Demo_conf from .cmi targets, not linked into galax.cma.
      - Removed shredded_load_update from .cmi targets, not used or
        linked into galax.cma.
      - Removed Galax_wrap from .cmi targets, not used or linked into
        galax.cma.
      - Made sure Stream_analysis_builtin and Stream_analysis are
        compiling, and linked-in properly. Added some basic
        'streaming' annotation to the core AST to do so. (Note: may
        have been intended to algebraic annotation, but never seemed
        to happen...)

December 1, 2009 - Jerome

  o Configuration:
      - Slight reorganization of the makefiles to allow for various
        sub-components to share most of the configuration, and differ
        on some of the target rules. Main galax configuration is now
        in ./config/Makefile.galax.common

December 1, 2009 - Jerome

  o Toplevel:
      - Fixed bug in command-line usage processing, which wrongly
        assumed it was always embedded in a galax executable, breaking
        user-generated executables with galax libraries in them.

December 1, 2009 - Jerome

  o Configuration:
      Fixed issues related to circular module dependencies being
      dropped during compilation of Galax.
        - Moved some signature and type definitions used in DXQ from
        server utilities to processing context.
        - Eliminated a rarely (never?) used dependency of the AST
        utils onto the AST pretty printer.

  o Code reorganization:
      Cleaned up some of the module dependencies and ensured properly
      staged compilation order for all the Caml modules.
        - Moved Logical_algebra_types from ./compilation to ./ast/Ast_logical_algebra_types.
        - Moved Alg_path_struct from ./projection to ./ast/Ast_path_struct.
        - Removed unused module in Streaming_ops (referencing the AST).
        - Removed unused module in Code_execute to (referencing
          Galax_server_util which is toplevel).

November 23, 2009 - Jerome

  o Toplevel:
      - Added schema parsing/mapping to main command-line tool:
          galax-mapschema [options]    --> galax xmlschema [options]

  o Config:
      - Fixed duplicate clean up of .depend files in some of the tools
        (removing Warning).

November 23, 2009 - Jerome

  o Documentation:
      - Cleaned up the documentation generation so it doesn't choke on
        nested URLs and properly generates the table of content.

November 23, 2009 - Jerome

  o Toplevel:
      - Started consolidating command-line executable under a single
        'galax' command. New commands are:
           galax-run [options]    --> galax xquery [options]
           galax-parse [options]  --> galax xml [options]
           galax-compile [options --> galax compile [options]

November 18, 2009 - Jerome

  o Examples:
      - Remove obsolete or incomplete examples directories.

November 12, 2009 - Jerome

  o Parsing:
      - Fixed error message properly pointing to DXQ option when parsing
        DXQ programs.

  o XQueryX embedding:
      - Re-enabled parsing/printing of XQueryX embedding of DXQ
        interfaces.

November 6, 2009 - Jerome

  o Toplevel:
      - Added build information to the version message.

November 6, 2009 - Jerome

  o Web site:
      - Updated the description of the compiler, reflecting the most
        recent architecture, along with pointers to corresponding
        specifications or published articles.

November 6, 2009 - Jerome

  o Web site:
      - Added pointers to DXQ articles.
      - Fixed some of the material in optimization.

  o Test suite:
      - Fixed issue with test configuration, not running the separate
        Galax test suite.

  o Caml:
      - Upgraded to O'Caml 3.11.1. Fixed some minor serialization
        issues for floating point numbers resulting from changes in
        the O'Caml standard library.

  o Datatypes:
      - Added proper interface for the datatype lexer.

October 29, 2009 - Jerome

  o Web Site:
      - Changes to the Web Site front pages for Sourceforge.

October 28, 2009 - Jerome

  o Config:
      - Minor cleanups of Makefiles. Notably missing clean targets in
        website and regression.

October 15, 2009 - Jerome

  o Minor fixes to top-level
  o Changes to Web site

August 18, 2009 - Jerome

  o Switched Galax to SVN server on sourceforge.
  o A number of bug fixes at various places (hard to locate without
    old CVS).

***** Tagged with galax-1-1 for GODI O'Caml release 3.10 *****

March 24, 2008 - Mary

  o Typing_call 

    - Fixed insidious bug in Typing_call in which static type checking
      for binary overloaded function call did not match on all
      possible pairs of types, thus excluding valid output types,
      which lead to type-based rewriting incorrectly converting the
      overloaded/polymorphic call to a monomorphic call.

March 21, 2008 - Jerome

  o Typing:
     - Fixed wrong signatures for op:is-same-node, op:node-*.

***** RE-Tagged with galax-1-0 for GODI O'Caml release 3.10 *****

March 12, 2008 - Jerome

  o Installation:
     - Now installing the extensions examples directory.
     - Added missing META.in.

  o Serialization:
     - Fixed a few issues with serialization (showing up while doing
       the tutorial).

  o Documentation:
     - Added table of contents back.
     - Pass on installation chapter. Added section on how to run the
       W3C XQuery test suite.
     - Removed the 'accessing and storing XML' part relating to Jungle
       and datamodel, empty or obsolete.
     - Added back chapter on APIs.
     - Shuffled the chapters in the reference part (alignment is
       renamed conformance and comes first, 'general' last).
     - Removed tutorial part on serialization as well-formed,
       buggy. Added text for more useful 'standard' mode.
     - Pass on tutorial chapter added a short tutorial section on
       printing compilation phases.

  o Testing:
     - Fixed the default config file to not point to the
       Galax-specific test suite. (for distribution).

February 21, 2008 - Jerome

  o Release:
     - Changed the status to 'Stable'

  o Documentation:
     - Pass on Chapter 1. 

February 21, 2008 - Jerome

  o Debian:
     - Folded in Stefano's patches.

  o Documentation:
     - Re-enabled documentation generation.
     - Fixed some formatting issues.

***** Tagged with galax-1-0 for GODI O'Caml release 3.10 *****

February 12, 2008 - Mary 

  o Ran XQuery Test Suite

  o Fixed the APIs 

January 29, 2008 - Mary [DXQ]

  o Galaxd:

    BUG ALERT:
    Added default configuration: Conf.set_materialize_tables Conf.Always;
    Changes Conf.materialize_tables_kind from Analysis (i.e., always
    do analysis) to Always, due to quadratic memory usage bug somewhere
    in Alg_path_structutil.path_sequences_with_disjoint_roots

  o Typing_expr:
    
    Bugs in typing Eval and Insert.

  o Webgui:

    Added new graph-drawing for <node> and (sub)<tree>
 
  o Top_util:

    load_graph raises error if named graph does not exist.

  o Procmod_phases:
  
    Permitted code selection phase to be optional

January 14, 2008 - Mary [DXQ]

  o Playing with Willow

December 5, 2007 - Mary

  o Java API:
    - Changes to get Java API to link
    
    TODO: align examples/galapi/java_api


November 26, 2007 - Mary 

  o C API:

    Makefile.galax: Had to add "-lrt" and CONF_C_LIBPCRE to C-API link
    opitons.  

November 16, 2007 - Mary [DXQ]

  o Changed syntax of box-closure expression to:
      for server P implement Q box { E }
    
    Assumption is that closure will be evaluated at some server P
    that implements interface Q, but we do not know the ultimate
    destination. 

    NB: Now that we ship arbitrary code in closure, we have even more
    motivation to distinguish between functions that are (1) location
    dependent and/or (2) updating.  We ue 'updating' modifier to
    describe both, which conflates meaning.  This will eventually bite
    us when we want to optimize an expression for a particular server.

  o Updated Willow example:
    - Changes friend for a domain based on latency
    
November 7, 2007 - Mary [DXQ]

  o Changed glx:gettime() to use Unix.gettimeofday() 

  o Fixed bug in normalization & code selection for 
    'for server P box { E }' to have same semantics as the remote
    execute expressions.

November 6, 2007 - Mary [DXQ]

  o Changed config files to use uniform graph format for neighbors and
    latencies.

    See examples/dxq/narada/latencies.xml & examples/dxq/willow/topology.xml

  o Webgui:
    Added "-t graph.xml" option to load initial neighbor topology from file. 

October 25, 2007 - Jerome

  o DXQ:
     - Fixed precedence conflicts in the parser for boxstuff
       expressions.

  o Anonymization:
     - Factored out occurrences of 'Galax' in strings into
       a single constant, to facilitate anonymization. (SIGMOD stuff).

***** Tagged with galax-0-9-9 for Debian Release to Stefano Zacchiroli <zack@bononia.it> *****

October 24, 2007 - Mary 

  o Documentation: 
    
    Woefully incomplete vis a vis Updates, XQueryP, APIs, etc. 

  o Code_selection:

    Fixed insidious bug in code selection for references to imported
    variable or imported function: Incorrectly returned
    code-selection-context of imported module instead of importing
    module.

  o configure script:

    Disabled C and Java APIs until we figure out what we want to do
    with them.

  o OCaml API examples:

    Aligned with new processing model 

October 23, 2007 - Jerome

  o Unicode:
     - Added support to change the location of Unicode character maps
       through an environment variables.

October 16, 2007 - Mary [DXQ]

  o Typing_expr:

    Added default typing rules for update expressions so that
    sub-expressions are also typed. 

  o Conf.mlp:

    XQueryP is containted in DXQ: Fixed is_xqueryp() to account for
    inclusion. 

  o Narada: 

    Initialized global variables to be non-empty, to avoid this bug:

    [TODO : There is a nasty stack overflow bug in Subtyping_gluskov.includes:
     Internal Error: Subtyping_gluskov.includes 
       small = element route of type xs:untyped, text?, element route of type xs:untyped (min 0) (max 0)
       big = attribute *  of type, (...rest of item()*...)
    ]

***** Tagged with galax-0-7-10 *****

October 15, 2007 - Mary 
  o Release:
    - Set config/Makefile.galax RELEASE number to 0.7.10

    Should this be 1.0...finally?!

October 15, 2007 - Mary [DXQ]

  o Added two new constructs to DXQ: boxing an expression in a closure
    and evaluating a boxed closure.  This expression:

      for server S box Expr 

    constructs a closure for Expr in the context of server
    implementation S.  The closure is an XML tree.  This expression:
    
      eval box Expr              

    unboxes the closure in Expr and evaluates the closure.  A
    rough definition of at-server-do (similarly, for-server-return) in terms
    of box and eval:

        at server S do E 
      === 
        let $v := for server S box E return
        at server S do { eval box $v }
 
    Semantics of these constructs still a bit fuzzy.  

  o Implementation required changes all the way from parser to code
    selection and plan serialization.

October 10, 2007 - Jerome

  o Join detection:
     - Disabled rewriting of "some" into "let" to facilitate join
       detection in a few specific cases (affects q10 in XMark).
     - Changed how XMark query 10 is written (manual inlining of two
       variables) to avoid issues with independance checking
       interacting with optimization.

September 20, 2007 - Jerome

  o Location hints:
     - Fixed schema_import.mlp to use the new location hint calls.

September 20, 2007 - Jerome

** Please run configure again due to changes to detect Camomile's version **

  o Galax Compilation:
     - Fixes to ./configure to support camomile's version.
     - Upgraded to Ocaml 3.10
       (changes in compilation due to new camlp4)
     - Added support for Camomile 0.7.* on top of Camomile 0.6.
       (changes due to new library names)
     - Now forcing use of Netsys, coming with latest ocamlnet.

  o Join optimization:
     - Another fix to the compiler, to properly clean-up selec
       clause for join detection.

Sept 25, 2007 - Mary [DXQ]

  o Grammar + Xquery_ast/Xquery_core_ast

    - Added optional "implements URI [at Expr]" clause to module declaration:

      "module" "namespace" NCNAME "=" URI ["implements" URI ["at" Expr]];

    - GLXTS : Galax Test Suite

      Added "Interfaces" test group for testing interfaces & modules
 
  o Processing_context
    - Location hints are completely managed in this module
    - Now distinguish between interface, module and schema location
      hints. 

Sept 19, 2007 - Mary [DXQ]

  o Grammar + Xquery_ast/Xquery_core_ast

    - Added global server declaration to prolog:
        "declare server NCNAME implements NCNAME at Expr;"
      aligns with :
        "let server NCNAME implement NCNAME at Expr return Expr"
    
      It doesn't match other declarations that define namespaces:
      e.g., "module namespace X = URI";

  o Namespace_names + many normalization modules

    Added support for interface and server prefixes:
    
    - Extended Namespace_names.prefix to distinguish between interface
      and server prefixes. 
      (Pre-requisite to implementing function shipping correctly.)

    - Fixed scoping of declare/let-server prefixes and added static
      checks that server-functions are actually defined in
      corresponding interface.

    Open questions/TODO
    ~~~~~~~~~~~~~~~~~~~
    x Should globally declared server in a module be imported into
      importing module?  Right now, it's not, but it probably should
      be. 

    x Norm_ident_top/expr are incomplete vis a vis declare-server and
      let-server.

  o DXQ Examples 
    Use new declare-server syntax

  o General clean-up of Web GUI 

    Webgui:
    - Flashplayer interface is now primary GUI
    - Excised all SVG code and event reporting from HTML interface
    - HTML interface only useful for quick access to servers'
      programs, debug logs, and sending queries
    - Still not very clean code, but there's a lot less to grapple
      with

September 15, 2007 - Jerome

  o Streaming:
     - Disabled stream inside tuple fields because of some serious
       problem with the streaming analysis...

September 13, 2007 - Jerome

  o Compilation:
    - Fixed bug in the compilation of where clauses, affecting some
      queries with conjunctive/disjunctive predicates.

  o Processing model:
    - Added a debug option for the compiler.

  o Namespaces:
    - Fixed expected results for Usecase NS, and regression tests.

September 12, 2007 - Mary

  o General clean-up of Web GUI 

    Webgui:
    - Flashplayer interface is now primary GUI
    - Excised all SVG code and event reporting from HTML interface
    - HTML interface only useful for quick access to servers'
      programs, debug logs, and sending queries
    - Still not very clean code, but there's a lot less to grapple
      with

September 10, 2007 - Mary

  o General clean-up of DXQ server code 

    Galax_server_util:
    - Contains server utility functions shared by Galax_server,
      Galaxd, Webgui: server logging, signatures for server functions,
      etc.

    Galax_server:
    - Consolidated GADS of replicated code: 
      queries from peers, queries from GUIs, error handling
    - Defines HTTP/TCP and UDP server functions
    - Builtin DXQ functions except those depending on
      Thread/Mutex/Condition modules
    - All calls to Browser/GUI isolated here 

    Galaxd:
    - Thread server definition and definitions of thread primitives
      exposed as builtin DXQ functions.

    - Mainline for server start-up: reads directories, parses files,
      etc., but does not handle query evaluation or call GUI.  All
      query evaluation and GUI I/O in Galax_server.

September 7, 2007 - Mary

  (Warning: NS,SGML,XMP usecases currently failing regression)

  o General clean-up of HTTP code

    Http : 

      All HTTP functions (client-side, server-side, and rpc-like) are
      now in http.ml.  This module is in a "hybrid" state: the HTTP
      get function is implemented using Http_client; all others are
      implemented using home-grown HTTP request code.

      Client-side functions only return HTTP message payloads (no
      status or headers) and an raise exception if any status other
      than 200/"OK" is received.

    Galax_http :

       Is replaced by Galax_url, which only handles URL decoding and
       user-installed schemes.

September 4, 2007 - Nicola

  o added a new function, glx:http-get-request() that (temporarily)
    implements HTTP get calls with full HTTP/1.1 features (using Netclient), 
    until we fix and unify the HTTP support.

August 30, 2007 - Jerome

  o Namespaces:
    - Added a couple of optimizations to the namespace resolution
      process.
    - Now avoiding serialization of namespace bindings that are
      already in scope.

August 29, 2007 - Nicola

  o Workflows:
    - Added support for HTML forms.

August 27, 2007 - Jerome

  o XQueryP:
    - Removed obsolete while normalization code.

August 27, 2007 - Jerome

  o XQueryP:
    - Fixed normalization of already normalized expressions with while
      loops.

August 27, 2007 - Jerome

   ** NOTE: Complete re-implementation of while loops in XQueryP. It

Changes
 -----------
| Changelog |
 -----------

May 22, 2019 - Max

 o Now it compiles. Compilation prerequisites: make, libpcre, opam, the following opam packages: pxp camomile camlidl num pcre.

May 8, 2019 - Max

 o Making Galax compile with Ocaml 4.07 and most recent library dependencies. I am not there yet.

February 1, 2012 - Jerome

 o Constructors:
    - Fixed bug with relative base URI in element constructors.

January 25, 2012 - Jerome

 o Parsing:
    - Fixed lexing issue with tokens after '/' that can only be in the
      operator state.

 o Constructors:
    - Fixed some error cases when xmlns/xml names are used in
      attribute or element constructors.

January 24, 2012 - Jerome

 o Parsing:
    - Added collation support in order-by clause.

January 23, 2012 - Jerome

 o Rewritings:
    - Fixed somewhat embarrassing bug on rewriting of overloaded
      functions, not properly checking for all cases in union types.
    - Fixed rewriting bug for overloaded function applied to anyURI,
      revealed by the previous fix.

January 20, 2012 - Jerome

 o Parsing:
    - Fixed a few regressions introduced by the latest change to the parser.

 o Streaming:
    - Fixed an issue with code-selection for global variables,
      streaming too aggressively.

 o Type-checking:
    - Fixed a bug in type intersection involving node().

January 19, 2012 - Jerome

 o Parsing:
    - Removing obolete rules leading to (harmless) shift/reduce.
    - Cleaned up some unused tokens and lexing rules.

January 18, 2012 - Jerome

 o F&O:
    - Final fix to fn:subsequence, issue with NaN or INF as parameter.
    - Fix to negative marker on fn:round resulting in -0 rather than 0.

January 17, 2012 - Jerome

 o F&O:
    - Fixed rounding issue with fn:subsequence.

January 10, 2012 - Jerome

 o Configuration:
    - Fixed bug in configuration generation introduced in previous patch.

 o F&O:
    - fn:doc-available returns false for the empty sequence, not empty.

January 10, 2012 - Jerome

 o Lexing:
    - Fixed lexical state issue arising from function types.
    - Added -print-states [on/off] option to the top-level tokenizer.

January 7, 2012 - Jerome

 o Configuration:
    - Fixed lexing tool script for MacOSX.

January 6, 2012 - Jerome

 o Lexing:
    - Added a hook to the tokenizer from the command line, returning
      the set of lexing tokens, along with the lexical state's stack.

      Usage:
        glx tokenize [files]

January 5, 2012 - Jerome

 o Lexing:
    - Removed Xmldecl_lexer, unused.

 o Parsing:
    - Computed processing instruction accepts no expression content.
    - Direct processing instruction constructor checks for space after
      NCName.
    - Computed processing instruction names, should make sure to cast
      to xs:NCName. (Note: cast to xs:NCName does not properly check
      for facet constraints at this point).

 o Overloaded functions:
    - Fixed overloading of unary arithmetic operators over the empty
      sequence.

 o Date and time:
    - Fixed rounding error when multiplying over a yearMonthDuration.

 o Functions and operators:
    - Added missing (new?) fn:doc-available, simply reusing fn:doc.

January 4, 2012 - Jerome

 o Documentation:
    - Some cleanup and basic update of the documentation.

 o Configuration:
    - Made the ./configure script more robust to variations in ocaml's
      version number.

 o Installation:
     - Fixed installed makefile for update usecases.

January 3, 2012 - Jerome

 o Parsing:
    - Fixed minor issue with new-line normalization.
    - Added strict conformance flag. Sample rule is no multiple
      expressions in the main module.
    - Fixed bug in parsing or processing instruction kind tests
      provided as a string, succeedings even when not a validNCName.
    - Fixed bug in parsing for attribute and schema-attribute node
      kinds.

December 21, 2011 - Jerome

  o Rewriting:
     - Fixed slightly too aggressive rewriting rules for
       fs:untyped-to-double, leading to errors in some cases.

December 20, 2011 - Jerome

  o Normalization:
     - Fixed odd bug on leading '/' and leading '//' not properly
       checking for document node.

  o Testing:
     - Simple switch of XQTS to v103
     - Added some missing test groups (some new namespace, errors, XML
       1.1 tests)

  o Parsing:
     - Fixed bug in parsing of default typeswitch clause with a variable.

  o XPath:
     - Fixed a couple of issues with reverse axis and
       following/preceding sibling (empty for attributes nodes).

  o Configuration:
     - Cleaned up a few wrong version references and links to Web site
       in the code.

December 16, 2011 - Jerome

  o Rewriting:
     - Reverted slightly too aggressive rewriting rule for casting,
       introducing a regression.
     - Consolidated E1[E2] rewrite, making sure they are correct and
       do not introduce regressions.

December 15, 2011 - Jerome

  o Rewriting:
     - Fixed some of the basic rewrites for E1[E2] when E2 is a
       numeric, leads to better access times for those cases.

December 8, 2011 - Jerome

  o Testing:
     - Fixed serialization of XML optimized plans during testing
     - Some fixes to the GLX tests catalog

December 8, 2011 - Jerome

  o Streaming:
     - Further cleanup of namespace resolution.

December 6, 2011 - Jerome

  o GIT:
     - Removed some files that shouldn't be tracked.
  o Streaming:
     - Slightly neater attribute resolution code.

December 6, 2011 - Jerome

  o Streaming:
     - Cleaned up some of the handling of namespace attributes.

December 6, 2011 - Jerome

  o Streaming:
     - Reduces namespace environment manipulation during stream
       resolution, resulting in much faster processing in some cases.

December 5, 2011 - Jerome

  o Build:
      - Now builds on ocaml 3.12.1+rc1
      - Upgraded all the major libraries:
      	   - findlib 1.2.7
           - pcre-ocaml 6.2.3
           - pxp 1.2.2
           - camomile 0.8.3

  o Data model:

base/debug.ml
(***********************************************************************)
(*                                                                     *)
(*                                 GALAX                               *)
(*                             XQuery Engine                           *)
(*                                                                     *)
(*  Copyright 2001-2007.                                               *)
(*  Distributed only by permission.                                    *)
(*                                                                     *)
(***********************************************************************)

(* $Id: debug.ml,v 1.10 2007/09/13 18:36:42 simeon Exp $ *)

(* Module: Debug
   Description:
     This module implements basic operations used for debugging.
*)

type debug_flag =
  | JoinDebug
  | TypeDebug
  | DefaultDebug
  | CompileDebug
  | StaticDebug
  | DxqDebug
  | MaterializationDebug

let debug_flag_of_string s =
  match s with
  | "join" -> JoinDebug
  | "typing" -> TypeDebug
  | "default" -> DefaultDebug
  | "compile" -> CompileDebug
  | "static" -> StaticDebug
  | "dxq" -> DxqDebug
  | "materialization" -> MaterializationDebug
  | _ -> raise Not_found

let string_of_flag f =
  match f with
  | JoinDebug -> " JOIN"
  | TypeDebug -> " TYPING"
  | DxqDebug -> " DXQ"
  | StaticDebug -> " STATIC"
  | CompileDebug -> " COMPILE"
  | MaterializationDebug -> "MATERIALIZATION"
  | DefaultDebug -> ""

let debug_flags = ref []

let print_debug flag msg =
  let fs = string_of_flag flag in
  Format.fprintf (!Conf.glx_err_formatter) "[DEBUG%s] %s@.%!" fs msg;
  Format.pp_print_flush (!Conf.glx_err_formatter) ()

let join_debug ()    = List.exists (fun x -> x = JoinDebug) !debug_flags
let dxq_debug ()     = List.exists (fun x -> x = DxqDebug) !debug_flags
let typing_debug ()  = List.exists (fun x -> x = TypeDebug) !debug_flags
let compile_debug ()  = List.exists (fun x -> x = CompileDebug) !debug_flags
let static_debug ()  = List.exists (fun x -> x = StaticDebug) !debug_flags
let materialization_debug ()  = List.exists (fun x -> x = MaterializationDebug) !debug_flags
let default_debug () = List.exists (fun x -> x = DefaultDebug) !debug_flags

let print_join_debug msg =
  if join_debug() then
  print_debug JoinDebug msg

let print_typing_debug msg =
  if typing_debug() then
  print_debug TypeDebug msg

let print_compile_debug msg =
  if compile_debug() then
  print_debug CompileDebug msg

let print_default_debug msg =
  if default_debug() then 
  print_debug DefaultDebug msg

let print_dxq_debug msg =
  if dxq_debug() then 
  print_debug DxqDebug msg

let print_static_debug msg =
  if static_debug() then 
  print_debug StaticDebug msg

let print_materialization_debug msg =
  if materialization_debug() then 
  print_debug MaterializationDebug msg

let sprintf_default_debug ff x =
  let s = Format.sprintf ff x in
  print_default_debug s

let set_debug df = debug_flags := df


